include/oneapi/dpl/pstl/utils.h

```diff
@@ -276,16 +263,32 @@ class __not_equal_value
 template <typename _Pred>
 class __transform_functor
 {
-    _Pred _M_pred;
+    mutable _Pred _M_pred;
 
   public:
-    explicit __transform_functor(_Pred __pred) : _M_pred(__pred) {}
+    explicit __transform_functor(_Pred __pred) : _M_pred(::std::move(__pred)) {}
 
     template <typename _Input1Type, typename _Input2Type, typename _OutputType>
     void
-    operator()(const _Input1Type& x, const _Input2Type& y, _OutputType& z) const
+    operator()(const _Input1Type& x, const _Input2Type& y, _OutputType&& output) const
+    {
+        __transform_impl(::std::forward<_OutputType>(output), x, y);
+    }
+
+    template <typename _InputType, typename _OutputType>
+    void
+    operator()(_InputType&& x, _OutputType&& output) const
```

Comment:
1) We do we have `_InputType&&` as input type while in the overload above we have `const _Input1Type& x` ?
2) What happens in case of rvalue type for output? How does a caller get a result?

Reply:
Let me try to answer to your question(2): please see my example at https://godbolt.org/z/cerxrvMT6

Reply:
With regards to (1).  I believe the original intention was to avoid enabling non-const lvalue references for inputs to the binary overload (which universal/ forwarding references would do).  

@rarutyun originally wrote the code to use const lvalue references explicitly for inputs for both binary and unary functors, but then changed it to support forwarding references for unary to support existing behavior. 
https://github.com/oneapi-src/oneDPL/pull/1233/commits/3e3c043e91f44a0a55b3f3ac2807a235707644ea

I think that was to support dereferenced `zip_iterators` forwarding along them as rvalues, but I am not sure if that is the reason (other than to not change the existing code).  Since this was changed intentionally, and aligns with the previous code, I believe it is very likely required.

Reply:
> but I am not sure if that is the reason

Yes, it was done to support dereferenced type for `zip_iterator`. It is a rvalue (like `tuple<ref, ref>`) which doesn't  match to lvalue type (`outputType&`).

Reply:
OK yes, since the dereference return is not a simple rvalue, but a proxy object, it wont just simply bind to `const T&` and work the way we would want.

However, with this in mind, I'm wondering how the current tests for `transform_binary_zip.pass` compile + pass with `zip_iterators` as inputs when the code only has const lvalue for binary inputs.  Would this same concept not prevent us from accepting `zip_iterator` inputs to binary transforms as it is currently written?

Reply:
Dan, It seems I don't catch what you mean...

Reply:
It seems now that the rvalue inputs for unary transform was not to support `zip_iterators` but rather to support those ill-formed operators in the `binary_search` algorithms (as fixed by your recent commit).  This makes more sense now how the tests pass using `zip_iterator` inputs.

Reply:
To go a little further and try to summarize (stating what might be obvious to others):

The derefenced `zip_iterators` return a rvalue ref of `tuple<const lvalue ref, const lvalue ref>` which will still bind to const lvalue ref input variables in `__transform_functor::operator()` .  However, for the output variable, an rvalue ref of `tuple<const lvalue ref, const lvalue ref>` with *not* bind to a non-const lvalue ref output variable.  This requires we use rvalue references for output variables, and forward them.  

Functionally, rvalue references would work for the input variables, but is more permissive than we would like to be, as it includes non-const lvalue references.  The implementation for the binary search algorithms was explicitly using a functor with lvalue references as input parameters, which was causing problems.

Reply:
Actually, `binary_search_impl.h` was reverted.
At the moment the compiling of `binary_search_impl.h` doesn't lead any issues.